alpha/alpha-fsm/src/main/java/org/apache/servicecomb/pack/alpha/fsm/repository/elasticsearch/ElasticsearchTransactionRepository.java

```diff
@@ -0,0 +1,204 @@
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import org.apache.servicecomb.pack.alpha.fsm.metrics.MetricsService;
+import org.apache.servicecomb.pack.alpha.fsm.repository.TransactionRepository;
+import org.apache.servicecomb.pack.alpha.fsm.repository.model.GloablTransaction;
+import org.apache.servicecomb.pack.alpha.fsm.repository.model.PagingGloablTransactions;
+import org.elasticsearch.action.search.SearchResponse;
+import org.elasticsearch.index.query.QueryBuilders;
+import org.elasticsearch.search.SearchHit;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.data.domain.PageRequest;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.elasticsearch.core.ElasticsearchTemplate;
+import org.springframework.data.elasticsearch.core.ScrolledPage;
+import org.springframework.data.elasticsearch.core.SearchResultMapper;
+import org.springframework.data.elasticsearch.core.aggregation.AggregatedPage;
+import org.springframework.data.elasticsearch.core.aggregation.impl.AggregatedPageImpl;
+import org.springframework.data.elasticsearch.core.query.GetQuery;
+import org.springframework.data.elasticsearch.core.query.IndexQuery;
+import org.springframework.data.elasticsearch.core.query.NativeSearchQueryBuilder;
+import org.springframework.data.elasticsearch.core.query.SearchQuery;
+
+public class ElasticsearchTransactionRepository implements TransactionRepository {
+
+  private static final Logger LOG = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+  public static final String INDEX_NAME = "alpha_global_transaction";
+  public static final String INDEX_TYPE = "alpha_global_transaction_type";
+  private static final long SCROLL_TIMEOUT = 3000;
+  private final ElasticsearchTemplate template;
+  private final MetricsService metricsService;
+  private final ObjectMapper mapper = new ObjectMapper();
+  private int batchSize;
+  private int batchSizeCounter;
+  private int refreshTime;
+  private final List<IndexQuery> queries = new ArrayList<>();
+  private final Boolean lock = true;
+
+  public ElasticsearchTransactionRepository(
+      ElasticsearchTemplate template, MetricsService metricsService, int batchSize,
+      int refreshTime) {
+    this.template = template;
+    this.metricsService = metricsService;
+    this.batchSize = batchSize;
+    this.refreshTime = refreshTime;
+    if (this.refreshTime > 0) {
+      new Thread(new RefreshTimer(), "elasticsearch-repository-refresh").start();
+    }
+    if (!this.template.indexExists(INDEX_NAME)) {
+      this.template.createIndex(INDEX_NAME);
+    }
+  }
+
+  @Override
+  public void send(GloablTransaction transaction) throws Exception {
+    long begin = System.currentTimeMillis();
+    queries.add(convert(transaction));
+    batchSizeCounter++;
+    metricsService.metrics().doRepositoryReceived();
+    if (batchSize == 0 || batchSizeCounter == batchSize) {
+      synchronized (lock) {
```

Comment:
I looked save method, I suppose the lock object here is to protect queries field from concurrent access.
I think here is a defect, line 82: `queries.add(convert(transaction))` is out of the lock scope so it's not thread safe.
I also suggest that you can make use of `synchronized(this)` here and change save method to `private synchronized void save(long begin)`, to make the code easier to understand.

Reply:
> I looked save method, I suppose the lock object here is to protect queries field from concurrent access.
> I think here is a defect, line 82: `queries.add(convert(transaction))` is out of the lock scope so it's not thread safe.
> I also suggest that you can make use of `synchronized(this)` here and change save method to `private synchronized void save(long begin)`, to make the code easier to understand.

No, the purpose of the lock is to avoid conflicts between timed writes and bulk writes.